### ISSUE TITLE
Validation images not matching ComfyUI

### ISSUE THREAD

**Issue by komninoschatzipapas** [bghira/SimpleTuner#409]
I'm curious as to why the validation images output are not matching the ones I'm getting from ComfyUI when using the same parameters.

![CleanShot 2024-05-17 at 19 51 33](https://github.com/bghira/SimpleTuner/assets/29049104/5bf66bce-a7a1-407e-92ef-08aeba5a3cff)

And then running `train_sdxl.py` with the following arguments and having added the line I mentioned on the other issue to perform validation on step 0:
```
  --validation_seed=42 \
  --validation_guidance=5 \
  --validation_guidance_rescale=0.0 \
  --validation_num_inference_steps=35 \
  --validation_negative_prompt="" \
  --validation_resolution=832x1216 \
  --validation_noise_scheduler="ddim"
```
Yields:
![step_0_car_(832, 1216)](https://github.com/bghira/SimpleTuner/assets/29049104/4cec3a77-6c1e-4a85-9699-dad2ce5e783b)

We perform inference using ComfyUI on our production service so ideally I would be able to have those two be in sync but I have not found a way to do this yet.

**Comment by bghira** [comment]
i've never used the validations pipeline that way, so i'm not sure that it would work at all. you can run the inference scripts in the toolkit to try directly via diffusers.

**Comment by komninoschatzipapas** [comment]
I used DDPM since there wasn't a DDIM script:
```python
# Use Pytorch 2!
import torch
from diffusers import (
    StableDiffusionPipeline,
    DiffusionPipeline,
    AutoencoderKL,
    UNet2DConditionModel,
    DDPMScheduler,
)
from transformers import CLIPTextModel

# Any model currently on Huggingface Hub.
# model_id = 'junglerally/digital-diffusion'
# model_id = 'ptx0/realism-engine'
# model_id = 'ptx0/artius_v21'
# model_id = 'ptx0/pseudo-journey'
model_id = "stabilityai/stable-diffusion-xl-base-1.0"
pipeline = DiffusionPipeline.from_pretrained(model_id)

# Optimize!
pipeline.unet = torch.compile(pipeline.unet)
scheduler = DDPMScheduler.from_pretrained(model_id, subfolder="scheduler")

# Remove this if you get an error.
torch.set_float32_matmul_precision("high")

pipeline.to("cuda")
prompts = {
    "car": "car",
}
for shortname, prompt in prompts.items():
    # old prompt: ''
    image = pipeline(
        prompt=prompt,
        negative_prompt="",
        num_inference_steps=35,
        generator=torch.Generator(device="cuda").manual_seed(42),
        width=832,
        height=1216,
        guidance_scale=5,
    ).images[0]
    image.save(f"./{shortname}_nobetas.png", format="PNG")
```
Yields:
![car_nobetas](https://github.com/bghira/SimpleTuner/assets/29049104/a7f60d54-d0da-4ade-b485-74483be4d546)

ComfyUI:
![CleanShot 2024-05-17 at 21 25 06](https://github.com/bghira/SimpleTuner/assets/29049104/7e62322e-42bc-44b6-bd95-de25dec789e1)

I'm not sure what I'm missing but its crazy how different these outputs are with seemingly the same model & parameters. My concern is that we're using ComfyUI for inference but judging our models based on the validation images generated by diffusers which don't seem to be generated in the same manner.

I'm going to close this issue as it's related to diffusers & ComfyUI but if you're able to give some input on this situation it would be greatly appreciated!

**Comment by bghira** [comment]
i haven't really played around with the base SDXL model in almost a year, it is possibly something in their upstream model config. i had to open a few issue reports so far for some of these problems, and a couple of them just remain perpetually unresolved

**Comment by bghira** [comment]
![image](https://github.com/bghira/SimpleTuner/assets/59658056/387bed02-508d-486f-8ab8-c55307bb5b2f)

here's my result, though I had to modify the script slightly for MPS, replacing `cuda` -> `mps`, and removing the `torch.compile()` because MPS does not support it

**Comment by bghira** [comment]
```py
# Use Pytorch 2!
import torch
from diffusers import (
    StableDiffusionPipeline,
    DiffusionPipeline,
    AutoencoderKL,
    UNet2DConditionModel,
    DDPMScheduler,
)
from transformers import CLIPTextModel

# Any model currently on Huggingface Hub.
# model_id = 'junglerally/digital-diffusion'
# model_id = 'ptx0/realism-engine'
# model_id = 'ptx0/artius_v21'
# model_id = 'ptx0/pseudo-journey'
model_id = "stabilityai/stable-diffusion-xl-base-1.0"
pipeline = DiffusionPipeline.from_pretrained(model_id)

# Optimize!
# pipeline.unet = torch.compile(pipeline.unet)
scheduler = DDPMScheduler.from_pretrained(model_id, subfolder="scheduler")

# Remove this if you get an error.
torch.set_float32_matmul_precision("high")

pipeline.to(dtype=torch.float16, device="mps")
prompts = {
    "car": "car",
}
for shortname, prompt in prompts.items():
    # old prompt: ''
    image = pipeline(
        prompt=prompt,
        negative_prompt="",
        num_inference_steps=35,
        generator=torch.Generator(device="mps").manual_seed(42),
        width=832,
        height=1216,
        guidance_scale=5,
    ).images[0]
    image.save(f"./{shortname}_nobetas.png", format="PNG")
```

**Comment by komninoschatzipapas** [comment]
Check https://github.com/comfyanonymous/ComfyUI/issues/3507

If you were to change `torch.Generator(device="mps").manual_seed(42)` to `torch.Generator(device="cpu").manual_seed(42)`, you should get the image I got from my ComfyUI flow above.

Is there a possibility we could integrate this into the codebase for generating the validation images?

**Comment by bghira** [comment]
no, because generator on CPU means all Tensors have to be copied from CPU during generation time. it even tells you this when you try it.

**Comment by bghira** [comment]
![image](https://github.com/bghira/SimpleTuner/assets/59658056/d2ceeef9-39f5-4076-a9d7-6b67a27a02bc)

using CPU seed on the mac doesn't seem to reproduce the original images you gave still, for what it's worth

**Comment by komninoschatzipapas** [comment]
I actually just tried it on my Studio M1 Ultra and I'm getting the same results as you.

So in order to be in sync with ComfyUI, you need to generate the seed on the CPU and be on a Linux system.

I'm relieved it's the seed that was different and not some other parameter that would be harder to adjust.

**Comment by bghira** [comment]
so basically no reason to switch since it's just a different seed being tested :-) you can probably sample using GPU seeds on ComfyUI as well.

**Comment by komninoschatzipapas** [comment]
> i haven't really played around with the base SDXL model in almost a year, it is possibly something in their upstream model config. i had to open a few issue reports so far for some of these problems, and a couple of them just remain perpetually unresolved
> 
> edit: i will try and reproduce this problem locally.

I'm curious what you're fine-tuning off. I found checkpoints like juggernaut overcook pretty easily so I generally go off base.

**Comment by bghira** [comment]
my own [model series](https://huggingface.co/collections/ptx0/terminus-xl-65451893a156b3b1d1456514) trained on a different noise schedule with the same u-net layout as SDXL, and same text encoder, but Made by Ollins' fp16 VAE.
